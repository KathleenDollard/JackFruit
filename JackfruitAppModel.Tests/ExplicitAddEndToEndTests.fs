namespace Generator

open Microsoft.CodeAnalysis;
open Xunit
open ApprovalTests.Reporters
open System
open Generator.Tests.UtilsForTests
open ApprovalTests
open System.Linq
open Generator.SourceGenerator
open Generator
open Generator.Tests.MapExplicitAddData
open Generator.ExplicitAdd

type ExplicitAddTestGenerator() =
    inherit Jackfruit.ExplicitAddGenerator()

    override _.CodeModelBuilder commandDef =
        BuildNewerCliCodeModel.OutputCommandWrapper commandDef

//[<Generator>]
type ExplicitAddEndToEndTests() =
    let RunCliModelGenerator (inputCompilation: Compilation) generator =
        RunGenerator generator inputCompilation

    let RunTest (data: Data) generator =
        let compilation = CreateCompilation data.CliCode
        let inputDiagnostics = compilation.GetDiagnostics()
        if not (inputDiagnostics.IsEmpty) then invalidOp "Input code is invalid"
        let (driver, outputCompilation, diagnostics) = RunCliModelGenerator compilation generator
        if not (inputDiagnostics.IsEmpty) then invalidOp "Generation failed"
        let newTree = outputCompilation.SyntaxTrees.Skip(1).SingleOrDefault()
        Assert.NotNull newTree
        Approvals.Verify(newTree.ToString())   

     
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``No commands``() =
        RunTest NoMapping (Jackfruit.ExplicitAddGenerator())
        
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``One simple command``() =
        RunTest OneMapping (Jackfruit.ExplicitAddGenerator())
    
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``Three commands``() =
        RunTest ThreeMappings (Jackfruit.ExplicitAddGenerator())

    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``No commands with test generator``() =
        RunTest NoMapping (ExplicitAddTestGenerator())
   
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``One simple command with test generator``() =
        RunTest OneMapping (ExplicitAddTestGenerator())
    
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``Three commands with test generator``() =
        RunTest ThreeMappings (ExplicitAddTestGenerator())
