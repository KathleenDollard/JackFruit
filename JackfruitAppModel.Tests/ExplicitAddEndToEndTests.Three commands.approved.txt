using System;
using System.CommandLine;
using System.CommandLine.Invocation;
using System.Threading.Tasks;
namespace CliDefinition
{
   
   public class OriginalSeriesCli : ICommandHandler
   {
      private Action<string, string, string> _operation {get; set;}
      public Command Command {get; set;}
      public OriginalSeriesCli(Action<string, string, string> operation)
      {
         _operation = operation;
         Command = new Command("OriginalSeries");
         kirkOption = new Option<string>("kirk");
         Command.Add(kirkOption);
         spockOption = new Option<string>("spock");
         Command.Add(spockOption);
         uhuraOption = new Option<string>("uhura");
         Command.Add(uhuraOption);
         Command.Handler = this;
      }
      public Option<string> kirkOption {get; set;}
      public string kirkOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(kirkOption);
      }
      public Option<string> spockOption {get; set;}
      public string spockOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(spockOption);
      }
      public Option<string> uhuraOption {get; set;}
      public string uhuraOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(uhuraOption);
      }
      public Task<int> InvokeAsync()
      {
         ._operation(.kirkOptionResult(context), .spockOptionResult(context), .uhuraOptionResult(context));
         return Task.FromResult(context.ExitCode);
      }
   }
   
   public class NextGenerationCli : ICommandHandler
   {
      private Action<string> _operation {get; set;}
      public Command Command {get; set;}
      public NextGenerationCli(Action<string> operation)
      {
         _operation = operation;
         Command = new Command("NextGeneration");
         picardOption = new Option<string>("picard");
         Command.Add(picardOption);
         Command.Handler = this;
      }
      public Option<string> picardOption {get; set;}
      public string picardOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(picardOption);
      }
      public Task<int> InvokeAsync()
      {
         ._operation(.picardOptionResult(context));
         return Task.FromResult(context.ExitCode);
      }
   }
   
   public class VoyagerCli : ICommandHandler
   {
      private Action<string, string, string, bool, string> _operation {get; set;}
      public Command Command {get; set;}
      public VoyagerCli(Action<string, string, string, bool, string> operation)
      {
         _operation = operation;
         Command = new Command("Voyager");
         janewayOption = new Option<string>("janeway");
         Command.Add(janewayOption);
         chakotayOption = new Option<string>("chakotay");
         Command.Add(chakotayOption);
         torresOption = new Option<string>("torres");
         Command.Add(torresOption);
         tuvokOption = new Option<bool>("tuvok");
         Command.Add(tuvokOption);
         sevenOfNineOption = new Option<string>("sevenOfNine");
         Command.Add(sevenOfNineOption);
         Command.Handler = this;
      }
      public Option<string> janewayOption {get; set;}
      public string janewayOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(janewayOption);
      }
      public Option<string> chakotayOption {get; set;}
      public string chakotayOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(chakotayOption);
      }
      public Option<string> torresOption {get; set;}
      public string torresOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(torresOption);
      }
      public Option<bool> tuvokOption {get; set;}
      public bool tuvokOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "bool">(tuvokOption);
      }
      public Option<string> sevenOfNineOption {get; set;}
      public string sevenOfNineOptionResult(InvocationContext context)
      {
         return context.ParseResult.GetValueForOption<SimpleNamedItem "string">(sevenOfNineOption);
      }
      public Task<int> InvokeAsync()
      {
         ._operation(.janewayOptionResult(context), .chakotayOptionResult(context), .torresOptionResult(context), .tuvokOptionResult(context), .sevenOfNineOptionResult(context));
         return Task.FromResult(context.ExitCode);
      }
   }
}
